crates/amaru-ledger/src/lib.rs

```diff
@@ -36,3 +36,212 @@ pub mod rules;
+                $test_directory,
+                "/data/transactions/preprod/",
+                $hash,
+                "/tx.cbor"
+            )))
+            .unwrap()
+        };
+        ($test_directory:literal, $hash:literal, $test_variant:literal) => {
+            cbor::decode::<KeepRaw<'_, MintedTransactionBody<'_>>>(include_bytes!(concat!(
+                $test_directory,
+                "/data/transactions/preprod/",
+                $hash,
+                "/",
+                $test_variant,
+                "/tx.cbor"
+            )))
+            .unwrap()
+        };
+    }
+
+    macro_rules! include_witness_set {
+        ($test_directory:literal, $hash:literal, $test_variant:literal) => {
+            from_cbor::<WitnessSet>(include_bytes!(concat!(
+                $test_directory,
+                "/data/transactions/preprod/",
+                $hash,
+                "/",
+                $test_variant,
+                "/witness.cbor"
+            )))
+            .unwrap()
+        };
+        ($test_directory:literal, $hash:literal) => {
+            from_cbor::<WitnessSet>(include_bytes!(concat!(
+                $test_directory,
+                "/data/transactions/preprod/",
+                $hash,
+                "/witness.cbor"
+            )))
+            .unwrap()
+        };
+    }
+
+    pub(crate) use include_transaction_body;
+    pub(crate) use include_witness_set;
+
+    // -----------------------------------------------------------------------------
+    // Tracing for Tests
+    // -----------------------------------------------------------------------------
+
+    #[derive(Clone)]
+    pub(crate) struct TestingTraceCollector {
+        pub lines: Arc<Mutex<Vec<String>>>,
+    }
+
+    impl TestingTraceCollector {
+        pub fn new() -> Self {
+            TestingTraceCollector {
+                lines: Arc::new(Mutex::new(Vec::new())),
+            }
+        }
+        pub fn clear(&mut self) {
+            self.lines.lock().unwrap().clear();
+        }
+    }
+
+    impl Write for TestingTraceCollector {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            if let Ok(s) = std::str::from_utf8(buf) {
+                for line in s.lines() {
+                    if !line.is_empty() {
+                        self.lines.lock().unwrap().push(line.to_string());
+                    }
+                }
+            }
+            Ok(buf.len())
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    pub(crate) fn with_tracing<F, R>(test_fn: F) -> R
+    where
+        F: FnOnce(&TestingTraceCollector) -> R,
+    {
+        let mut collector = TestingTraceCollector::new();
+        collector.clear();
+        let collector_clone = collector.clone();
+        let format_layer = fmt::layer()
+            .with_writer(move || collector_clone.clone())
```

Comment:
What is going on here 😅

Reply:
I forgot I didn't want to commit this 😅 

I was having issues with `with_writer` and ownership (as you can see)